main/src/addins/MacPlatform/MainToolbar/SearchBar.cs

```diff
@@ -259,20 +275,59 @@ bool IsFirstResponderOfWindow (NSWindow window)
 			return false;
 		}
 
-		bool SendKeyPressed (Xwt.KeyEventArgs kargs)
+		void SendKeyPressed (Xwt.KeyEventArgs kargs)
 		{
 			if (KeyPressed != null)
 				KeyPressed (this, kargs);
+		}
+
+		internal event EventHandler<SearchEntryCommandArgs> PerformCommand;
+
+		[Export ("control:textView:doCommandBySelector:")]
+		bool CommandBySelector (NSControl control, NSTextField field, ObjCRuntime.Selector sel)
+		{
+			SearchPopupCommand command;
+
+			switch (sel.Name) {
+			case "moveDown:": // down arrow
+				command = SearchPopupCommand.NextItem;
+				break;
+
+			case "moveUp:": // up arrow
+				command = SearchPopupCommand.PreviousItem;
+				break;
+
+			case "scrollPageDown:": // page down
+			case "moveToEndOfDocument:": // cmd+down arrow
+				command = SearchPopupCommand.NextCategory;
+				break;
+
+			case "scrollPageUp:": // page up
+			case "moveToBeginningOfDocument:": // cmd+up arrow
+				command = SearchPopupCommand.PreviousCategory;
+				break;
+
+			case "insertNewline:": // Return
+				command = SearchPopupCommand.Activate;
+				break;
+
+			case "cancelOperation:": // Escape
+				command = SearchPopupCommand.Cancel;
+				break;
+
+			default:
+				return false;
+			}
 
-			return kargs.Handled;
+			var args = new SearchEntryCommandArgs (command);
+			PerformCommand?.Invoke (this, args);
```

Comment:
This prevents creating the event args unless the handler is set.

```suggestion
			PerformCommand?.Invoke (this, new SearchEntryCommandArgs (command));
```